### ISSUE TITLE
Improvement suggestions for matplotlib.Animation.save('video.mp4')

### ISSUE THREAD

**Issue by Zulko** [matplotlib/matplotlib#3865]
TLDR : Maybe it would be possible to speed up 2x matplotlib's video writing by sending directly a raw RGB input to the writer (ffmpeg, avconv). Also, Matplotlib.Animation's default encoding is bad.

Hi there,

I compared writing a video using Matplotlib's animation module, with writing the same Matplotlib video using my library MoviePy instead of Animation.save().

See [here](https://github.com/Zulko/moviepy_matplotlib) for details and results.

It turns out that MoviePy produces an animation of better quality (Matplotlib's has
JPEG-compression-style artefacts) and much lighter (95k for MoviePy, 456k for Matplotlib).
MoviePy is also much faster (generating the animation takes 6 seconds with MoviePy, against 12s
for Matplotlib).

For the quality/size, it may be a matter of codecs or whatever (I really can't find what), I'm sure it can be tuned.

For the speed, however, it may be due to the fact that Matplotlib converts the frame into some compressed image format (I believe it is PNG by default) and sends them to the writer (ffmpeg in my case), which then must decode it. MoviePy sends the raw RGB snapshot of the matplotlib canvas directly to ffmpeg (no encoding/decoding needed).

What do you think of that ?

**Comment by videobrain** [comment]
Just to add an alternative view about possible use cases: I would like to animate charts, but need high quality video output to combine that with other video material. Compression will happen with the final product at the end of the pipeline, but the output of Matplotlib should be very high quality so the material can be still used in the following video transformation steps.

**Comment by tacaswell** [comment]
Closing this as streaming writer exist in the code base for ffmpeg and libav (ex http://matplotlib.org/api/animation_api.html?highlight=animation#matplotlib.animation.FFMpegWriter).  mpl passes the full-resolution RBGA buffer via a pipe to the encoder.  Exactly how that is compressed is up to the underlying software (and you can pass through what ever cl args you want).
